include/eve/product_type.hpp

```diff
@@ -69,4 +69,122 @@ namespace eve
 
   template<typename Type>
   inline constexpr bool supports_ordering_v = supports_ordering<Type>::value;
+
+   //================================================================================================
+  //! @addtogroup struct
+  //! @{
+  //!   @struct supports_like
+  //!   @concept like
+  //!
+  //!   `like<Wrapper, T>` is a concept that describes a wrapper for which most of the functionality
+  //!   for T should be applicable. T is always like T.
+  //!
+  //!   `like< Wrapper, T>` is true if Wrapper defines a typedef `using is_like = T` or
+  //!   specialises `eve::supports_like`.
+  //!
+  //!   Example: `eve::wide<T>` is `like<T>`.
+  //!
+  //!   This is used to work with custom structs in eve, see `oop` section in `examples`.
+  //! @}
+  //================================================================================================
+
+
+  template<typename Wrapper, typename Self> struct supports_like : std::false_type {};
+
+  template <typename Wrapper, typename Self>
+  requires requires(Wrapper) { typename Wrapper::is_like; }
+  struct supports_like<Wrapper, Self> :
+  std::bool_constant<std::same_as<Self, typename std::remove_cvref_t<Wrapper>::is_like>>
+  {};
+
+  template<typename Wrapper, typename Self>
+  concept like = std::same_as<std::remove_cvref_t<Wrapper>, Self> ||
+                 supports_like<std::remove_cvref_t<Wrapper>, Self>::value;
+
+  //================================================================================================
+  //! @addtogroup struct
+  //! @{
+  //!   @struct product_type_base<Self, ...Fields>
+  //!   @brief  a helper CRTP base to declare user defined types easier
+  //!
+  //!   It will generate a member wise equality.
+  //!   It will also generate a lexicographical order for your type, unless !supports_ordering_v<Self>
+  //!
+  //!   It generated the not mutating operator version from a mutating one (like operator+ from operator+=),
+  //!   however does not go further than that (like operator-= b from operator+= and operator-)
+  //!   since simd has an instruction for `-` and you want `operator-=` to be declared separately.
+  //!
+  //! @}
+  template <typename Self, typename ... Fields>
+  struct product_type_base;
```

Comment:
I guess the `Fields...` is because we're are building a type with shortcuts and not adapting an existing type ?

Reply:
Not a big fan of base in the name but this is basically bikeshedding

Reply:
I'm open to suggestions for names.
...Fields is to specify the types of fields, I don't get your comment tbh.